Operational Monitoring.

emtwo opened the pull request:
This PR is in progress. Here are the items that need to be done for it to be ready for review:
- [x] init & query file generation
- [x] output to bq destination table
- [x] allow running from bqetl cli
- [x] dimensions/probes/slugs/branches should be pulled in based on external files

emtwo commented:
Associated airflow PR: https://github.com/mozilla/telemetry-airflow/pull/1322

emtwo commented:
Right now the dry-run fails because I am referencing a `mozfun` function that I added in this PR. I can add it in a separate PR if that's better

scholtzan commented on bigquery_etl/operational_monitoring/templates/query.sql:
```diff
@@ -0,0 +1,94 @@
+    `{{source}}`
+  WHERE
+    DATE(submission_timestamp) >= DATE_SUB(@submission_date, INTERVAL 30 DAY)
+  GROUP BY
+    1,
+    2,
+    3,
+    4,
+    5,
+    6
```
It looks like `dimensions` could be an arbitrary set of dimensions. Could this break the `GROUP BY` here (e.g. if `dimensions` happen to contain 100 different dimensions)?

emtwo replied:
oh good catch, thank you

scholtzan commented on bigquery_etl/operational_monitoring/templates/query.sql:
```diff
@@ -0,0 +1,94 @@
+    {{ "," if not loop.last else "" }}
+  {% endfor %}
+)
+GROUP BY
+  1,
+  2,
+  3,
+  4,
+  5,
+  6
```
Same concern as above query. It might also be better to use the column names here instead to avoid confusion

scholtzan commented on sql/moz-fx-data-shared-prod/operational_monitoring/bug-1660366-pref-ongoing-fission-nightly-experiment-nightly-83-100/query.sql:
```diff
@@ -0,0 +1,89 @@
+-- Generated via bigquery_etl.operational_monitoring
```
I guess generated files should not be checked in into the repository. Looks like the DAG in https://github.com/mozilla/telemetry-airflow/pull/1322/files would generate the queries before running them

scholtzan commented on sql/moz-fx-data-shared-prod/operational_monitoring/bug-1660366-pref-ongoing-fission-nightly-experiment-nightly-83-100/init.sql:
```diff
@@ -0,0 +1,30 @@
+CREATE TABLE IF NOT EXISTS
```
This should probably also be removed before merging

scholtzan commented on bigquery_etl/operational_monitoring/__init__.py:
```diff
@@ -0,0 +1,177 @@
+    bq_client = bigquery.Client(project=PROD_PROJECT)
+
+    for slug in os.listdir(os.path.join(OUTPUT_DIR, dataset)):
+        destination_table = f"{project}.{dataset}.{slug}"
```
For the `slug` part, generally the naming convention for tables is not to use dashes but to use underscores instead. This is how we do it in jetstream: https://github.com/mozilla/jetstream/blob/29efddb0825678b55504ccc07bfe86c4363bdad9/jetstream/__init__.py#L22

scholtzan commented on bigquery_etl/operational_monitoring/__init__.py:
```diff
@@ -0,0 +1,177 @@
+def _run_sql(project, dataset, submission_date):
+    bq_client = bigquery.Client(project=PROD_PROJECT)
+
+    for slug in os.listdir(os.path.join(OUTPUT_DIR, dataset)):
```
This should probably also run in parallel.

scholtzan commented on bigquery_etl/operational_monitoring/__init__.py:
```diff
@@ -0,0 +1,177 @@
+    dimensions, _ = _download_json_file(project, bucket, "dimensions.json")
+
+    # Iterating over all defined operational monitoring projects
+    for blob in bucket.list_blobs(prefix=PROJECTS_FOLDER):
```
I haven't checked how long the generation takes, but if it takes significant time we might also run this in parallel

emtwo commented:
@scholtzan I addressed your comments. I didn't parallelize the query generation yet though (but parallelized the run). It seems to only take a couple seconds or less and right now we're only starting with 1 project anyways so we can add it later.
